### ISSUE TITLE
Cannot provide Collection Definitions in referenced helper assemblies

### ISSUE THREAD

**Issue by analogrelay** [xunit/xunit#1541]
Repro Project: https://github.com/anurse/XunitCollectionsFromReferences

Steps:
1. Clone the repo
1. `dotnet restore` and `dotnet build` at the root
1. `cd` to `src\TestProject` and `dotnet test`

Expected behavior: Everything works

Actual behavior:

```
Test run for C:\Users\anurse\Code\anurse\XunitCollectionsFromReferences\src\TestProject\bin\Debug\netcoreapp2.0\TestProject.dll(.NETCoreApp,Version=v2.0)
Microsoft (R) Test Execution Command Line Tool Version 15.5.0-preview-20171009-10
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
[xUnit.net 00:00:00.3836234]   Discovering: TestProject
[xUnit.net 00:00:00.4522010]   Discovered:  TestProject
[xUnit.net 00:00:00.4946865]   Starting:    TestProject
[xUnit.net 00:00:00.5938625]     TestProject.TestClassUsingHelperCollection.Test1 [FAIL]
[xUnit.net 00:00:00.5954055]       A test class may only define a single public constructor.
[xUnit.net 00:00:00.6227238]   Finished:    TestProject
Failed   TestProject.TestClassUsingHelperCollection.Test1
Error Message:
 A test class may only define a single public constructor.


Total tests: 2. Passed: 1. Failed: 1. Skipped: 0.
```

What I'm trying to do is define a Test Collection in a helper assembly and use a base class and the inherited `CollectionAttribute` to assign tests to that collection. I'm using Fixtures to demonstrate this in my repro, but only to make clear the problem I'm seeing, which is that the Collection Definition is never loaded up. In my actual scenario, what I want to use is `[CollectionDefinition(DisableTestParallelization=true)]` (see https://github.com/aspnet/Testing/pull/294 for an example of how I've hacked it for now using shared-source packages 😭) to allow all tests of a certain type to run sequentially and isolated from the other tests. This was surprising to me as the Collection behavior does work cross-assembly (i.e. all types inheriting from my base class are indeed getting put into the same collection), but the `CollectionDefinitionAttribute` is completely ignored and it took debugging through some Xunit code to actually realize why.

Scanning all assemblies for `CollectionDefinitionAttribute`s might not be a great plan though. @pakrym and I discussed a bit and considered the possibility of scanning the assembly containing the base class which provides the `CollectionAttribute`, when using an inherited `CollectionAttribute`.

**Comment by bradwilson** [comment]
This has come up a couple times before (#374, #409, #710) and we have historically decided not to fix this issue, for performance reasons, as you alluded to in your final paragraph. The "base class" is a new twist on the problem, but I'm not 100% sure whether that edge case makes the usage any more understandable for people (i.e., it works when you have a base class but doesn't when you don't...?)

**Comment by anurse** [comment]
Yeah, I recognize the performance problem. One way we handle reducing assembly scanning in MVC is by only scanning assemblies that reference the assembly containing the base/attribute class in question, though that requires a little help from components like `Microsoft.Extensions.DependencyModel` (which I know you're a huge fan of :trollface:).

I may try to play with this a bit and see how the performance is, because it's pretty useful (particularly in our case). Perhaps an assembly-level attribute marking an assembly as a candidate for this scanning?

**Comment by bradwilson** [comment]
Yeah, unfortunately M.E.DM only works on .NET Core. There is no equivalent for desktop CLR or UWP.

An assembly-level attribute would be nice if it could be automatically attached when the user adds the `xunit.core` package. We already have props/targets for that package, so adding a .cs/.vb/.fs file shouldn't be too difficult, assuming we have a way of knowing what language they're using from inside the project file...

**Comment by anurse** [comment]
> An assembly-level attribute would be nice if it could be automatically attached when the user adds the xunit.core package.

Hmm, that is an interesting idea. Might be worth having a play with.

**Comment by bradwilson** [comment]
Hello,

I'm closing this issue because I believe it has been resolved through the use of `Type`-based collection references in v3, which allow collection references across projects.
